lib/grammar.js

```diff
@@ -1,823 +1,243 @@
-          }
+  const allTokens = [
+    Whitespace,
+    OpenToken,
+    StartLinePosition,
+    EndLinePosition,
+    OpenTokenPosition,
+    CaratPosition,
+    Only,
+    Identifier,
+    Period,
+    OpenParens,
+    CloseParens,
+  ];
+
+  const lexer = new Lexer(allTokens);
+
+  class GrammarParser extends Parser {
+    constructor(input) {
+      var col = 1;  // eslint-disable-line no-var
+      super(input, allTokens, false /* isErrorRecoveryEnabled */);
+      const $ = this;
+      const openTokenLength = '// '.length;
+
+      this.root = $.RULE('assertion', () => {
+        $.OPTION(() => $.SUBRULE($.capturedSpaces));
+        $.SUBRULE($.openToken);
+        const positions = $.SUBRULE($.positions);
+        const scopes = $.SUBRULE($.scopes);
+        return [positions, scopes];
+      });
+
+      this.capturedSpaces = $.RULE('capturedSpaces', () => {
+        const spaces = $.CONSUME(Whitespace);
+        col = col + spaces.image.length;
+        return spaces;
+      });
+      this.spaces = $.RULE('spaces', () => $.CONSUME(Whitespace));
+
+      this.openToken = $.RULE('openToken', () => {
+        const token = $.CONSUME(OpenToken);
+        col = col + token.image.length;
+      });
+
+      this.startPosition = $.RULE('startPosition', () => {
+        $.CONSUME(StartLinePosition);
+        return [0];
+      });
+
+      this.endPosition = $.RULE('endPosition', () => {
+        $.CONSUME(EndLinePosition);
+        return [-1];
+      });
+
+      this.openPosition = $.RULE('openPosition', () => {
+        $.CONSUME(OpenTokenPosition);
+        return [col - openTokenLength];
+      });
+
+      this.caratPosition = $.RULE('caratPosition', () => {
+        $.OPTION(() => $.SUBRULE($.capturedSpaces));
+        const carat = $.CONSUME(CaratPosition);
+        col = col + carat.image.length;
+        return col - 1;
+      });
+
+      this.caratPositions = $.RULE('caratPositions', () => {
+        const positions = [];
+        $.AT_LEAST_ONE(isCaratLookahead, () => {
+          const pos = $.SUBRULE(this.caratPosition);
+          positions.push(pos);
+        }, 'caratPosition');
+        return positions;
+      });
+
+      this.positions = $.RULE('positions', () => $.OR([
+        { ALT: () => $.SUBRULE($.startPosition) },
+        { ALT: () => $.SUBRULE($.endPosition) },
+        { ALT: () => $.SUBRULE($.openPosition) },
+        { ALT: () => $.SUBRULE($.caratPositions) },
+      ]));
+
+
+      this.only = $.RULE('only', () => {
+        $.CONSUME(Only);
+        return '=';
+      });
+
+      this.modifier = $.RULE('modifier', () => $.SUBRULE($.only));
+
+      this.scopeSuffix = $.RULE('scopeSuffix', () => {
+        const period = $.CONSUME(Period);
+        const identifier = $.CONSUME(Identifier);
+        return period.image + identifier.image;
+      });
+
+      this.scope = $.RULE('scope', () => {
+        const identifier = $.CONSUME(Identifier);
+        let scope = identifier.image;
+        $.MANY(() => {
```

Comment:
This looks like  a fully qualified name (a.b.c.d)
Its is a reoccurring pattern which has a connivence parsing rule for.
**AT_LEAST_ONE_SEP**

Example:
https://github.com/SAP/chevrotain/blob/1e338c98c29e215c5d4576f4e312f884dd7119aa/test/parse/recognizer_spec.ts#L147

unless you explicitly want the scope and scopeSuffix to be separate grammar rules?